Add --file Option to ioc_scanner.py and Clarify Help Messages

mcdonnnj opened the pull request:
# <!--- Provide a general summary of your changes in the Title above -->

## 🗣 Description

This PR adds a `--file` option to `ioc_scanner.py` if it is run in a stand-alone style. It also clarifies the help for the `--file` and `--stdin` options in `ioc_scan_cli.py` to (hopefully) make it more clear what they do.
<!--- Describe your changes in detail -->

## 💭 Motivation and Context

The misunderstanding of script functionality in #7 led to a problem and a use-case being discovered. This PR thus seeks to close #8 and #10 since they are related. This will hopefully improve the end user experience.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## 🧪 Testing

Automated tests pass and I tested the stand-alone usage changes on the testing server I used for #11 . Without a `--target` option in `ioc_scanner.py` it is not feasible to add test coverage because it will scan from the root of the filesystem.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## 🚥 Types of Changes

<!--- What types of changes does your code introduce? -->
<!--- Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (causes existing functionality to change)

## ✅ Checklist

<!--- Go over all the following points, and put an `x` in all the
boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask.
We're here to help! -->

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

mcdonnnj committed:
Add --file option to ioc_scanner.py for stand-alone usage

This adds a --file option to allow the ioc_scanner.py script to use a file on
the remote host as a source for IOC hashes. The README has also been updated to
provide an example of this usage.

mcdonnnj committed:
Update help for --file and --stdin options

As was noted in issue #7 the usage for those options did not clearly describe
what they did. This resulted in end user confusion or misunderstanding about
functionality.

jsf9k commented on README.md:
````diff
@@ -70,6 +70,15 @@ ansible --inventory=hosts-file cool-servers --module-name=script \
+```console
+ansible --inventory=hosts-file cool-servers --module-name=script \
+--args="src/ioc_scan/ioc_scanner.py --file hash_file.txt" --become \
+--ask-become-pass --user="ian.kilmister"
````
Lemmy!

felddy replied:
🤘

jsf9k reviewed (approved):
This looks great to me!

hillaryj reviewed (approved):
🥇

felddy commented on src/ioc_scan/ioc_scanner.py:
```diff
@@ -158,4 +158,20 @@ def main(blob=None, root="/"):
 
 if __name__ == "__main__":
-    sys.exit(main())
+    import argparse
+
+    parser = argparse.ArgumentParser(
+        description="Indicators of compromise (IoC) scanning tool."
+    )
+    parser.add_argument(
+        "-f", "--file", dest="hashfile", help="Get IOC hashes from specified file."
+    )
+    args = parser.parse_args()
+
+    if args.hashes_file:
+        logging.debug("Reading hashes from %s", args.hashfile)
+        with open(args.hashfile) as f:
+            hashblob = f.read()
+        sys.exit(main(hashblob))
+    else:
+        sys.exit(main())
```
Why are we putting this here and not in `main()`?

felddy replied:
Nevermind.  I had to review this more in Atom.   I remember `ioc_scan_cli.py` now.

hillaryj replied:
Generally, you put the ArgParser in the section for `if __name__ == "__main__"` so it only gets called if the script is run from the command line and thus has command-line arguments. Then it checks and dispatches validated arguments to the actual program functionality.

hillaryj replied:
Looks like someone is having a case of the Mondays... besides me. :) Gonna mark this as resolved.

felddy replied:
I think I can come up with a counter-argument with regards to code coverage testing when things live outside of main().   

Which reminds me...

mcdonnnj replied:
Typical functionality is `ioc_scanner.main()` being called from `ioc_scan_cli.main()`. So I thought the limited argument parsing should only be used when the script is being run in standalone mode.

felddy reviewed (changes requested):
Looks good.  Have a question about the argument parsing location.

felddy reviewed (approved):
Looks good.

felddy reviewed (changes requested):
I think we need some test coverage for the new parser.

mcdonnnj commented:
> 
> 
> I think we need some test coverage for the new parser.

The problem I ran into with that is that without a `--target` option it will default to scanning the root of the filesystem. I don't see how to get reliable, specific output under that circumstance.

dav3r commented on src/ioc_scan/ioc_scanner.py:
```diff
@@ -158,4 +158,20 @@ def main(blob=None, root="/"):
+    )
+    args = parser.parse_args()
+
+    if args.hashes_file:
```
Shouldn't this be `if args.hashfile:`?

mcdonnnj replied:
Ah, glad you caught that mistake. I used that name when originally creating the parser, but realized it didn't align with how things are named in `ioc_scan_cli.py`. I missed that one reference when changing the name to be consistent, but it should be fixed in 6058935.

dav3r reviewed (approved):
I'm good with this once you fix the error I noted. 👍

mcdonnnj committed:
Fix incorrect variable reference

I used "hashes_file" when originally writing this functionality. After testing
I realized that did not match how `ioc_scan_cli.py` referred to the same
argument. When I renamed the variable to store and reference, this poor, lone
reference was missed.

mcdonnnj committed:
Refactor for standalone mode and add standalone mode testing

Testing against the argument parsing was requested and a small refactor was the
only way I saw feasible to do so. This allows the main() function in
ioc_scanner.py to only perform command line argument parsing before calling the
core logic function. This also allowed me to add the pyfakefs package to make
testing this new functionality feasible.

mcdonnnj committed:
Flesh out all tests to match test_scan_stdin

Make testing consistent by adding the extra test in test_scan_stdin to all the
test functions.
Clarify the language in some of the test assertions.

dav3r commented on tests/test_ioc_scan.py:
```diff
@@ -111,7 +116,57 @@ def test_scan_stdin(capsys):
+
+def test_ioc_scanner_standalone_file(capsys, test_fs):
+    """Test running the scanner in standalone mode with an input target file."""
+    test_hash = "69630e4574ec6798239b091cda43dca0"
```
A small improvement here would be to change this `test_hash` value to illustrate that the detected IOC is in fact coming from `testblob.txt` and not from `test_hash`.

mcdonnnj replied:
Just so I'm on the same page, do you mean the value or the variable name?

dav3r replied:
I'm talking about changing the value of the `test_hash`, since it's currently the same as the EICAR test hash in `testblob.txt` that is being detected.

mcdonnnj replied:
`test_hash` is being used to verify the output of the tool. It needs to be the same value as what's in `testblob.txt` because that's the hash that should be detected when scanning:
https://github.com/cisagov/ioc-scanner/blob/ce0a0b8ceb5abf424723138288824ad278b4b2da/tests/test_ioc_scan.py#L167-L172

mcdonnnj replied:
The variable name could (should?) be changed to more clearly represent how it's being used.

dav3r replied:
Ah- you are right, I am wrong.  What I'm looking for here is a way to show that the IOCs we're searching for are actually coming from the input file, rather than from the [defaults here](https://github.com/cisagov/ioc-scanner/blob/improvement/add_file_option_to_stand-alone_mode/src/ioc_scan/ioc_scanner.py#L23L50).

mcdonnnj replied:
That makes sense and would be a good addition.

mcdonnnj committed:
Add additional testing for standalone mode

Check the logging output to ensure that we are reading from a file for hashes
when an argument is passed. Also check to make sure we are not when no argument
is passed.

mcdonnnj committed:
DRY testing and add new hash to testblob.txt

Put the commonly used values like the test file for --file and repeatedly used
hash values as globals. Added the tests/targets/lorem.txt file's SHA256 to the
testblob.txt file. Tests for providing a file as a source for hashes will now
check that lorem.txt and its SHA256 are in the results. This makes sure we can
confirm that it is only detected if we provide a file.

dav3r reviewed (approved):
Looks great- thanks for making those late-breaking testing changes!  🧪🧪 ⭐

mcdonnnj committed:
Update test that was missed and add new test for ioc_scan_cli

One test was not switched to checking for the lorem.txt target when it should
have been. I have also added a test that leverages pyfakefs to test the module
with default options.

dav3r reviewed (approved):
Nice testing additions! 👍

jsf9k reviewed (approved):
Nice use of `pyfakefs`.

mcdonnnj committed:
Merge branch 'develop' into improvement/add_file_option_to_stand-alone_mode

mcdonnnj committed:
Let isort do its thang on the branch's code

felddy commented on tests/test_ioc_scan.py:
```diff
@@ -23,6 +23,18 @@
+
+
+@pytest.fixture
+def test_fs(fs):
```
Nice!

felddy reviewed (approved):
Great job.  Thanks for making this shiny! 🌟
